src/sentry/region_to_control/producer.py

```diff
@@ -1,91 +1,178 @@
+from __future__ import annotations
+
 import atexit
 import dataclasses
-from typing import Any, Optional
+from abc import abstractmethod
+from typing import TYPE_CHECKING, Any, Optional
 
 from arroyo import Topic
 from arroyo.backends.kafka import KafkaPayload, KafkaProducer
 from django.conf import settings
 
-from sentry.models import AuditLogEntry, UserIP
 from sentry.region_to_control.messages import RegionToControlMessage, UserIpEvent
+from sentry.services.hybrid_cloud import InterfaceWithLifecycle, silo_mode_delegation
 from sentry.silo import SiloMode
 from sentry.utils import json, kafka_config
 
+if TYPE_CHECKING:
+    from sentry.models import AuditLogEntry
 
-def produce_user_ip(event: UserIpEvent):
-    if _should_produce_to_kafka():
-        write_region_to_control_message(
-            dataclasses.asdict(RegionToControlMessage(user_ip_event=event)), sync=False
-        )
-    else:
-        UserIP.objects.create_or_update(
-            user_id=event.user_id, ip_address=event.ip_address, values=dataclasses.asdict(event)
+
+class RegionToControlMessagingService(InterfaceWithLifecycle):
+    @abstractmethod
+    def write_region_to_control_message(self, payload: Any, sync: bool):
+        pass
+
+
+_empty_headers = list()
+
+
+class KafkaBackedRegionToControlMessageService(RegionToControlMessagingService):
+    _publisher: Optional[KafkaProducer]
+
+    def __init__(self):
+        self._publisher = None
+
+    def close(self):
+        if self._publisher is not None:
+            self._publisher.close()
+
+    def write_region_to_control_message(self, payload: Any, sync: bool):
+        future = self.get_region_to_control_producer().produce(
```

Comment:
Given that `RegionToControl` is in the class names you could skip including it here as well. Could do `write_message()` and `get_producer().produce()` 🤷

Reply:
yes, that's definitely true.  I can make another pass here soon on that.